src/kernel/mutex.c

```diff
@@ -14,70 +14,168 @@
+   queue implementation in Linux by Peter Zijlstra, et al. */
+
+static inline cpuinfo mutex_get_next(mutex m, cpuinfo ci, cpuinfo prev)
 {
-    assert(!ctx->waiting_on);
-    assert(!m->turn);
-    m->turn = ctx;
+    do {
+        /* If we're removing ci from the list and it is the last item queued,
+           try to rewind tail to point to prev.
+
+           If we're unlocking ci and at the end of the list, then this is just
+           a simple removal. */
+        if (m->mcs_tail == ci &&
+            compare_and_swap_64((u64*)&m->mcs_tail, u64_from_pointer(ci),
+                                u64_from_pointer(prev)))
+            return 0;
+
+        /* If a spin waiter has latched onto us, exchange its next with
+           null. This is necessary to keep step A spinning until the node
+           being removed gets a new prev link. */
+        if (ci->mcs_next) {
+            cpuinfo next = pointer_from_u64(atomic_swap_64((u64*)&ci->mcs_next, 0));
+            if (next)
+                return next;
+        }
+        mutex_pause();
+    } while (1);
 }
 
-static inline boolean mutex_cas_take(mutex m, context ctx)
+static inline void mcs_unlock(mutex m, cpuinfo ci)
 {
-    boolean acquired = compare_and_swap_64((u64*)&m->waiters_tail, 0,
-                                           u64_from_pointer(ctx));
-    if (acquired)
-        mutex_acquired(m, ctx);
-    return acquired;
+    /* cover race between m->mcs_tail swap and write to ci->mcs_next */
+    if (compare_and_swap_64((u64*)&m->mcs_tail, u64_from_pointer(ci), 0))
+        return;        /* no successor */
+
+    /* atomic_swap_64 uses __ATOMIC_SEQ_CST; this is safe, though release
+       semantics would be sufficient here */
+    cpuinfo next = pointer_from_u64(atomic_swap_64((u64*)&ci->mcs_next, 0));
+    if (!next)
+        next = mutex_get_next(m, ci, 0);
+
+    if (next)
+        next->mcs_waiting = false;
 }
 
 static inline boolean mutex_lock_internal(mutex m, boolean wait)
 {
     cpuinfo ci = current_cpu();
     context ctx = get_current_context(ci);
 
-    mutex_debug("cpu %d, mutex %p, wait %d, ra %p\n", ci->id, m, wait,
-                __builtin_return_address(0));
+    mutex_debug("mutex %p, wait %d, ra %p\n", m, wait, __builtin_return_address(0));
     mutex_debug("   ctx %p, turn %p\n", ctx, m->turn);
 
     /* not preemptable (could become option on allocate) */
     if (m->turn == ctx)
         halt("%s: lock already held - cpu %d, mutex %p, ctx %p, ra %p\n", __func__,
              ci->id, m, ctx, __builtin_return_address(0));
 
-    assert(!ctx->next_waiter);
-    if (!wait)
-        return mutex_cas_take(m, ctx);
+    assert(!ctx->waiting_on);
+    if (!wait) {
+        boolean acquired = false;
+        if (compare_and_swap_64((u64*)&m->mcs_tail, u64_from_pointer(ci), 0)) {
+            acquired = compare_and_swap_64((u64*)&m->turn, 0, u64_from_pointer(ctx));
+            mcs_unlock(m, ci);
```

Comment:
As far as I can see, at this stage we haven't touched the MCS lock at all, so mcs_tail will never equal `ci`, and the mcs_unlock() call here is unnecessary.

Reply:
This is a bug; the CAS of m->mcs_tail is supposed to exchange 0 -> ci, not ci -> 0. The idea here is that we don't want a try lock to jump ahead of other MCS spinners and grab the mutex out-of-turn, so it first needs to grab the (uncontended) MCS lock before trying to grab the mutex (and it unlocks MCS either way).